File: app/src/main/java/valiev/timur/tapfight/game/GameActivity.java
package valiev.timur.tapfight.game;

import android.content.Intent;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.view.View;
import android.widget.ProgressBar;
import android.widget.TextView;

import java.util.HashMap;

import valiev.timur.tapfight.GameOverActivity;
import valiev.timur.tapfight.Player;
import valiev.timur.tapfight.R;

public class GameActivity extends AppCompatActivity implements GameView {

    private HashMap<Player, TextView> scoreViews = new HashMap<>();
    private ProgressBar progressBar;

    private GamePresenter presenter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_game);

        scoreViews.put(Player.P1, (TextView) findViewById(R.id.p1Score));
        scoreViews.put(Player.P2, (TextView) findViewById(R.id.p2Score));

        progressBar = findViewById(R.id.progressBar);
        progressBar.setProgress(0);
        progressBar.setMax(PresenterImpl.FIGHT_TIME);

        presenter = new PresenterImpl(this, new ModelImpl());
    }

    @Override
    public void updatePlayerScore(Player player, int score) {
        scoreViews.get(player).setText(Integer.toString(score));
    }

    @Override
    public void endGame(int p1Score, int p2Score) {
        Intent intent = new Intent(this, GameOverActivity.class);
        intent.putExtra("P1_SCORE", p1Score);
        intent.putExtra("P2_SCORE", p2Score);
        startActivity(intent);
        finish();
    }

    @Override
    public void updateProgressBar(int progress) {
        progressBar.setProgress(progress);
    }

    public void onTapBtn(View view) {
        switch (view.getId()) {
            case R.id.p1Button:
                presenter.tapBtn(Player.P1);
                break;
            case R.id.p2Button:
                presenter.tapBtn(Player.P2);
                break;
        }
    }
}

File: app/src/main/java/valiev/timur/tapfight/presentation/game/view/GameActivity.kt
package valiev.timur.tapfight.presentation.game.view

import android.animation.ObjectAnimator
import android.content.Intent
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.view.View
import android.view.animation.LinearInterpolator
import io.reactivex.Observable
import io.reactivex.subjects.BehaviorSubject
import kotlinx.android.synthetic.main.activity_game.*


import valiev.timur.tapfight.presentation.gameOver.GameOverActivity
import valiev.timur.tapfight.R
import valiev.timur.tapfight.domain.entities.PlayerId
import valiev.timur.tapfight.presentation.game.presenter.GamePresenter
import valiev.timur.tapfight.repositories.impl.preferences.GamePreferencesDAO
import valiev.timur.tapfight.repositories.preferences.GamePreferencesRepository

class GameActivity : AppCompatActivity(), GameView {

    private var presenter = GamePresenter(this)

    private val preferences: GamePreferencesRepository = GamePreferencesDAO.instance

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_game)

        presenter.startGame()
    }

    override fun updatePlayerScore(player: PlayerId, score: Int) {
        when (player) {
            PlayerId.P1 -> p1_score
            PlayerId.P2 -> p2_score
        }.text = score.toString()
    }

    override fun startTimerAnimation() {
        ObjectAnimator.ofInt(
                progress_game_time,
                "progress",
                progress_game_time.progress,
                progress_game_time.max
        ).apply {
            interpolator = LinearInterpolator()
            duration = preferences.fightTimeSec * 1000
            start()
        }
    }

    override fun endGame(p1Score: Int, p2Score: Int) {
        val intent = Intent(this, GameOverActivity::class.java)
        intent.putExtra("P1_SCORE", p1Score)
        intent.putExtra("P2_SCORE", p2Score)
        startActivity(intent)
        finish()
    }


    private val tapBehaviorSubject: BehaviorSubject<PlayerId> = BehaviorSubject.create()

    override fun getTapObservable(): Observable<PlayerId> = tapBehaviorSubject

    fun onTapBtn(view: View) {
        tapBehaviorSubject.onNext(
                when (view.id) {
                    R.id.button_p1 -> PlayerId.P1
                    R.id.button_p2 -> PlayerId.P2
                    else -> PlayerId.P1
                }
        )
    }

    override fun onDestroy() {
        presenter.destroy()
        super.onDestroy()
    }
}
